### ISSUE TITLE
Enhanced request to retrieve member number.

### ISSUE THREAD

**Issue by meteos77** [textbrowser/biblioteq#331]
Explanation:
When copying the member number (in the members manager), this manager asks if we want to save the changes when we've just selected the number to copy to the clipboard.

Request:
Can you make a small button to copy the number to the clipboard so as not to touch the field containing the number.
The same style as for sending a letter at the end of the field.

**Comment by textbrowser** [comment]
You've already requested this in a discussion. I would create meaningful member names. Instead of m9999, saturn_saturn. Automatic completion is another option. I'm not adding copy-this and copy-that actions. They're not natural. Context-menus are nice but they're not accessible. More buttons require more descriptions.

**Comment by textbrowser** [comment]
In Glitch, I added enhanced context menus. For example, double clicking an item activates a context menu. The menus are also present as separate dialogs as well as being available through traditional mechanisms. Many interfaces don't have left buttons or one has to press special keys. Context-menus don't exist in Android. The entire concept needs to be moved away from. It's archaic and broken.

**Comment by textbrowser** [comment]
In a browser, it's impossible to activate the context menu with modifier keys and the primary mouse key. One has to click the right or the secondary button or whatever the other button is. It's stupid and not portable.  Other considerations would be timed context menus. Press for 5 seconds a menu appears. But that's much work because pressing and dragging or pressing and selecting. Mice interfaces are dumb and the frameworks have remained in the 1980s.

**Comment by textbrowser** [comment]
If one is in a windowed environment, one is in a windowed environment where things must be accessible either through a lot of clear buttons or smart mechanisms. But things must be presented as being present while also being hidden because of limitations. A hint that a context menu exists and is available in some environments while being available in all environments. So I spent a lot of time doing that in Glitch but I'm not doing it elsewhere because those projects are complete. Look forward and not in the past. So anyway, I'm not interesting in copy-to-this-and-that. Other paradigms must be presented.

**Comment by textbrowser** [comment]
My design is based on the harsh restrictions of Android. Although I don't work in Android, I think of how I could be doing this in Android. If I cannot do it in Android because of how Android is, I implement a solution for Android because that solution will work with other systems. And of course there are exceptions there too.

TL;DR

Less of hidden or weird options. More smart options and proper data. m999 -> saturn, etc.

**Comment by meteos77** [comment]
For the librarian, I prefer the person's first and last names to find the right account.

the mxxxx is a data for the computer, moreover nobody remembers its number from where the problem of research.

I go through the member manager to get the number because I have to copy/paste it into BA/Emprunt.

android compatible suggestion?

If I type the beginning of the name in BA/Emprunt and BQ suggests the list starting with A
A1 first name
A2 first name
A3 first name 
this simplifies the process enormously.

**Comment by textbrowser** [comment]
These models do not work well with thousands of accounts. A single letter may return 10,000 entries. Five letters may create 500 entries.  The solution requires two queries. First query determines if the results are normal. Second query produces the results for an input. When you visit a library, the computer does not guess your identity by displaying information of other members. The concept of a filter is nonsensical. That is, one does not reserve items for a patron which is unknown and becomes known by a filtration process. Type in the entire account.

**Comment by textbrowser** [comment]
Will be reluctantly included in the next release.

**Comment by meteos77** [comment]
Since its creation, you have made the right choices to improve BiblioteQ and we are very satisfied with the software.

Please don't make changes reluctantly, as this indicates that they are not the right ones.

As a user, it's my job to ask for changes based on my feedback, 

As a developer, you have to be aware of technical constraints to get the best software, and you've explained why.

I understand the need to keep developments compatible with Android and the best technical choices to have the best possible software over time.


The possible future: in 4 years, we'll have a computer with android OS ... thank you textbrowser for thinking of us :-)

**Comment by textbrowser** [comment]
Searching for a parameter from a list of many parameters is meaningful. It's similar to a dictionary or a phone book.

In Batch Activities, one presumably reserves items for a known patron. Perhaps the patron is unknown or partially known and this is the essence of your request. Again, this is abnormal behavior. Modern and archaic library systems would not list the patrons during reservation; that is, the patron would be identified and the librarian would not require a query to discover this patron through filtration. A library is a catalog and catalogs are **required** to be meaningful.

**Comment by meteos77** [comment]
In theory, the member should have a card with his or her barcode number on it and come along to borrow.

In this situation, no problem.

- We don't have a card yet :-(
- The showerhead needs to be reconfigured to accept codes other than ISBN ... where's the notice anyway?

Our migration to BiblioteQ has gone well, but we need to make a few small adjustments for it to work better.

- Card creation is one of them.

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/assets/10701156/8f29ce70-415f-4b51-9d3f-2d05907d5fda

**Comment by meteos77** [comment]
I see you've implemented the automatic-completion idea -> that's great thanks.

Before implementing it, I have a few questions:

To choose the right information to put in the identifier instead of the mxxxxxxxxxx automatically created by BQ 
We can put a string of 16 characters.
People give their name (not enough because of homonym), so first name in addition.

Problem of field size

This produces nom_prenom, which is more than 16 characters in some cases.

- if I automatically shorten it to 16 characters, it will look strange because even if only 1 letter is missing? Jocelyn instead of Jocelyne Christian / Christiane ; Joseph / Josephine ... (Male or Female first name)
- Organization names with more than 16 characters "CRECHE PETITE ENFANCE" , "CRECHE SAINT-LAURENT" ... CRECHE SAINT-LAURENT", surnames > 16.


- Can [-_] be used?
- Do you recommend lower-case letters only, not [A-Z]?
- The identifier is used for loans, so what restrictions [a-z] and [0-9] or accented characters are allowed [àéèûôêïî]?


Once in place, this will make borrowing easier because you won't need to have the member manager open (more space on the screen).

Thanks again for the progress

**Comment by textbrowser** [comment]
First initial + last name. There is a reminder to widen the field. Fifteen letters for the last name are sufficient.

Pluto Saturn -> psaturn. 16 letters allow for 26 ^ 16 possibilities in English or 43608742899428874059776 which is a huge number.

**Comment by textbrowser** [comment]
Why are you asking a question about valid characters when you can test the input yourself?

**Comment by textbrowser** [comment]
The completer ignores case and filters by a contains algorithm. Ends-with and starts-with are restrictive. Lengthy account names are annoying. When I connect to PQ, I don't want to type 16 characters for an account, followed by a lengthy password. Keep account names brief.

**Comment by textbrowser** [comment]
You're overthinking this. Will you have 20,000 patrons? No. And if you do? They can create nicknames. They don't need to be connected to their actual names.

**Comment by textbrowser** [comment]
Back in the day, I was considering 64 characters or infinitely-many characters. But I ignored it because 26 ^ 16 is huge. Actually it's more like (26 + 10 + 1) ^ 26 and more in other languages. So anyway, nicknames.

**Comment by meteos77** [comment]
I'm not thinking too much, I just want to get off to a good start and your thoughts are useful, for example the example of the name + 1st letter of the first name (pluto saturne becomes psaturne).

We need an identifier that can be obtained without going into the member manager, otherwise there's no point in evolving.

**Comment by textbrowser** [comment]
First initial + middle initial + last name. If the last name is too long, first 14 characters.

Or first initial + '.' + last name. Or first name + '_' + last name. If there is the potential for duplicates, consider birthdays. If those are also duplicates, consider nicknames.

**Comment by meteos77** [comment]
We're having a meeting on Tuesday, so I'll be able to propose the evolution with different examples.

**Comment by meteos77** [comment]
This works well
One comment: when creating a new user, either enter the whole name or restart BiblioteQ so that the new name with only a few letters is taken into account in the search.

Once restarted, it works perfectly. Could you please improve this?

**Comment by textbrowser** [comment]
?

**Comment by textbrowser** [comment]
Added test1 and I'm able to search for it in BA.

**Comment by textbrowser** [comment]
It's a Qt-ism. The model / database. Anyway.

**Comment by textbrowser** [comment]
Resolved yesterday or the day before. A new widget allows for refreshing of the model as the model is not populated nor modified if the database is.

**Comment by meteos77** [comment]
The functionality works well with the little extra button, thank you.

I made some statistics with 1st letter of first name + last name (without accented letter)

member found in how many letters
63.5% with 2 letters only
30.5% with 3 letters
2.5% with 4 letters

and 3.5% homonymous, so use the member manager.

The result is extraordinary!
